docs/vc-spec.md

```diff
@@ -1,6 +1,163 @@
+};
+type Icrc21ErrorInfo = record { description : text; error_code : nat64 };
+type IssueCredentialError = variant {
+    Internal : text;
+    SignatureNotFound;
+    InvalidIdAlias;
+    UnauthorizedSubject;
+    UnknownSubject;
+};
+type IssuedCredentialData = record { vc_jws : text };
+type PrepareCredentialRequest = record {
+    signed_id_alias : SignedIdAlias;
+    credential_spec : CredentialSpec;
+};
+type PrepareCredentialResponse = variant {
+    Ok : PreparedCredentialData;
+    Err : IssueCredentialError;
+};
+type PreparedCredentialData = record { vc_jwt : opt text };
+type SignedIdAlias = record {
+    credential_jws : text;
+    id_alias : principal;
+    id_dapp : principal;
+};
+service : {
+    consent_message : (Icrc21ConsentMessageRequest) -> (
+    Icrc21ConsentMessageResponse,
+    );
+    get_credential : (GetCredentialRequest) -> (GetCredentialResponse) query;
+    prepare_credential : (PrepareCredentialRequest) -> (
+    PrepareCredentialResponse,
+    );
+}
+```
+
+### 1: Consent Message
+
+In the attribute sharing flow a user must approve the issuance of a verifiable
+credential by an issuer, and this happens by approving a human-readable consent message from the issuer.
+See [ICRC-21](https://github.com/dfinity/wg-identity-authentication/blob/main/topics/icrc_21_consent_msg.md) for more detail.
+
+Identity provider requests the consent message via `Icrc21ConsentMessageRequest`,
+and upon successful response displays the consent message from `Icrc21ConsentInfo` to the user.
+
+### 2: Prepare Credential
+
+Preparation of a credential involves checking the validity of the request,
+and upon success, preparation of the actual credential requested by the user.
+
+```candid
+service : {
+  prepare_credential : (PrepareCredentialRequest) -> (PrepareCredentialResponse, );
+};
+
+type PrepareCredentialRequest = record {
+    signed_id_alias : SignedIdAlias;
+    credential_spec : CredentialSpec;
+};
+
+type SignedIdAlias = record {
+    credential_jws : text;
+    id_alias : principal;
+    id_dapp : principal;
+};
+
+type PrepareCredentialRequest = record {
+    signed_id_alias : SignedIdAlias;
+    credential_spec : CredentialSpec;
+};
+
+type PrepareCredentialResponse = variant {
+    Ok : PreparedCredentialData;
+    Err : IssueCredentialError;
+};
+
+type PreparedCredentialData = record { vc_jwt : opt text };
+```
+
+Specifically, the issuer checks via `prepared_id_alias.credential_jws` that user identified via `id_dapp` on the issuer
+side can be identified using `id_alias` for the purpose of attribute sharing, and that the credential
+described by `credential_spec` does apply to the user.  When these checks are successful, the issuer
+prepares and returns the credential in JWT format.
+
+**NOTE**: `prepare_credential`-service is needed only for some issuers, e.g. those that use [canister signatures](https://internetcomputer.org/docs/current/references/ic-interface-spec/#canister-signatures)
```

Comment:
This is backwards, no? `get_credential` is only needed if you do canister signatures. If you do e.g. threshold ECDSA then you can return the signed credential already in `prepare_credential` because you don't need the data certificate. However, it is _not_ possible to get a threshold ECDSA signature in `get_credential` because it is a query call.

In general, to keep the API consistent, I would make all the issuers implement both methods and just have them split the work in different ways.

Reply:
I believe we had the same intentions in mind, but the description was somewhat misleading.  Changed now, PTAL.